src/python/pants/base/exiter.py

```diff
@@ -72,8 +57,13 @@ def exit(self, result=0, msg=None, out=None):
     """
     if msg:
       if PY2:
-        msg = msg.encode('utf-8')  # sys.stderr expects bytes in Py2, unicode in Py3
-      print(msg, file=out or sys.stderr)
+        # sys.stderr expects bytes in Py2, unicode in Py3
+        msg = msg.encode('utf-8')
+      out = out or sys.stderr
+      if not is_fileobj_definitely_closed(out):
+        print(msg, file=out)
+        # NB: Ensure we write everything out in case it's not an unbuffered stream like stderr.
```

Comment:
I think that `f.close()` implicitly flushes.

Reply:
If we close stderr here we may see an error when invoking pants as a pex because the pex executor will call `sys.stderr.flush()` directly (haven't checked yet, but that was the error we saw in our internal CI). pex writing directly to stderr is something that I believe could be configurable and which would be extremely useful. maybe. i'll make a followup ticket.

Reply:
As in, it calls that in the body of a `finally:` after the pex exits, and then raises an exception if stderr is closed.

Reply:
The code is now all the same, but wrapped in a try/catch which calls `logger.exception(e)`. If it is almost definitely better to close it here, I can do it -- just not sure if there's a reason to do it over the flush alone. I've left a detailed TODO above the flush() call about this.